numpy/lib/shape_base.py

```diff
@@ -16,9 +16,192 @@
+    dest_dims = list(range(axis)) + [None] + list(range(axis+ins_ndim, indices.ndim))
+
+    # build a fancy index, consisting of orthogonal aranges, with the
+    # requested index inserted at the right location
+    fancy_index = []
+    for dim, n in zip(dest_dims, arr.shape):
+        if dim is None:
+            fancy_index.append(indices)
+        else:
+            ind_shape = shape_ones[:dim] + (-1,) + shape_ones[dim+1:]
+            fancy_index.append(_nx.arange(n).reshape(ind_shape))
+
+    return tuple(fancy_index)
+
+
+def take_along_axis(arr, indices, axis):
+    """
+    Take the elements described by `indices` along each 1-D slice of the given
+    `axis`, matching up subspaces of arr and indices.
+
+    This function can be used to index with the result of `argsort`, `argmax`,
+    and other `arg` functions.
+
+    For each `i...`, `j...`, `k...` in turn (representing a series of indices),
+    where the number of indices in `i...` is equal to `axis`, this computes::
+
+        out[i..., j..., k...] = arr[i..., indices[i..., j..., k...], k...]
+
+    Or equivalently (where `...` alone is the builtin `Ellipsis`):
+
+        out[i..., ..., k...] = arr[i..., :, k...][indices[i..., ..., k...]]
+
+    .. versionadded:: 1.13.0
+
+    Parameters
+    ----------
+    arr: array_like (Ni..., M, Nk...)
+        source array
+    indices: array_like (Ni..., Nj..., Nk...)
+        indices to take along each 1d slice of `arr`
+    axis: int
+        the axis to take 1d slices along
+
+    Returns
+    -------
+    out: ndarray (A..., K..., B...)
+        The indexed result.
+
+        out[a..., k..., b...] = arr[a..., indices[a..., k..., b...], b...]
+
+
+    See Also
+    --------
+    take : Take along an axis without matching up subspaces
+
+    Examples
+    --------
+
+    For this sample array
+
+    >>> a = np.array([[10, 30, 20], [60, 40, 50]])
+
+    We can sort either by using sort directly, or argsort and this function
+
+    >>> np.sort(a, axis=1)
+    array([[10, 20, 30],
+           [40, 50, 60]])
+    >>> ai = np.argsort(a, axis=1); ai
+    array([[0, 2, 1],
+           [1, 2, 0]], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
+    array([[10, 20, 30],
+           [40, 50, 60]])
+
+    The same works for max and min:
+
+    >>> np.max(a, axis=1)
+    array([30, 60])
+    >>> ai = np.argmax(a, axis=1); ai
+    array([1, 0], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
+    array([30, 60])
+
+    If we want to get the max and min at the same time, we can stack the
+    indices first
+
+    >>> ai_min = np.argmin(a, axis=1)
+    >>> ai_max = np.argmax(a, axis=1)
+    >>> ai = np.stack([ai_min, ai_max], axis=axis); ai
+    array([[0, 1],
+           [1, 0]], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
+    array([[10, 30],
+           [40, 60]])
+    """
+    # normalize inputs
+    arr = asanyarray(arr)
+    indices = asanyarray(indices)
+    if axis is None:
```

Comment:
Could you add support for arr.ndim > indices.ndim?
It's a single line here (+ unit test + docs):
```
indices = indices.reshape((1, ) * (arr.ndim - indices.ndim) + indices.shape)
```

Use case / example:
```
x = np.array([[5, 3, 2, 8, 1],
              [0, 7, 1, 3, 2]])
# Completely arbitrary y = f(x0, x1, ..., xn), embarassingly parallel along axis=-1
# Here we only have x0, but we could have more.
y = x.sum(axis=0)
# Sort the x's, moving the ones that cause the smallest y's to the left
take_along_axis(x, np.argsort(y))

Reply:
I don't think your use-case is well-motivated. A more explicit way to achieve that would be:
```
y = x.sum(axis=0, keepdims=True)
take_along_axis(x, np.argsort(y, axis=1), axis=1)
```

Reply:
The whole point is that with the one-liner addition f(x) can add or remove axes at will (some manual broadcasting required if it _replaces_ axes, which however happens automatically if you e.g. wrap this in ``xarray.apply_ufunc``).

Reply:
Note that in the comments above, we decide that perhaps it's best to not allow any case other than `indices.ndim == arr.ndim`, since there's no obvious right choice.

`take_along_axis` only really makes sense if you endeavor to keep all your axes aligned. `xarray` can probably solve that by axis names alone, but in numpy you need to indicate that by axis position. Therefore, you can't afford to let your axes collapse, and have numpy guess which one you lost: in your case, you're advocating for it to guess the left-most one should be reinserted - but this is only the case because you did `sum(axis=0)`.